### PR TITLE
Added user authentication, dashboard api.

### PR DESCRIPTION
**Pull request by nick-ser**

### CODE REVIEW

File: Smartling.API/Authentication/AuthApiUser.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        public override AuthResponse Refresh(string refreshToken)
+        {
+            var command = new RefreshCommand() { refreshToken = refreshToken };
+            var request = PrepareJsonPostRequest(ApiGatewayUrl + AUTH_API_V2_REFRESH_USER, command, string.Empty);
+            var jsonResponse = GetResponse(request);
+            var refreshResponse = JsonConvert.DeserializeObject<AuthResponseWrapper>(jsonResponse);
+
+            return refreshResponse.response;
+        }
```
Comment (PavelLoparev) [inline Smartling.API/Authentication/AuthApiUser.cs]: Why did you copy-pasted all the class contents? I expected that you will extend the existing class and override uri only.

File: Smartling.API/Authentication/AuthApiUser.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+﻿using Newtonsoft.Json;
+
+namespace Smartling.Api.Authentication
+{
+    public class AuthApiUser : AuthApiClient
+    {
+        private const string AUTH_API_V2_AUTHENTICATE_USER = "/auth-api/v2/authenticate/user";
+        private const string AUTH_API_V2_REFRESH_USER = "/auth-api/v2/authenticate/refresh/user";
```
Comment (PavelLoparev) [inline Smartling.API/Authentication/AuthApiUser.cs]: Did you try to refresh token with this url? I think refresh url should not be updated.

File: Smartling.API/Authentication/AuthApiClient.cs [hunk 3]
```diff
@@ -5,35 +5,36 @@ namespace Smartling.Api.Authentication
...
+          var request = PrepareJsonPostRequest(ApiGatewayUrl + AUTH_API_V2_REFRESH, command, string.Empty);
+          var jsonResponse = GetResponse(request);
+          var refreshResponse = JsonConvert.DeserializeObject<AuthResponseWrapper>(jsonResponse);
 
-      return refreshResponse.response;
-    }
-  }
+          return refreshResponse.response;
+        }
+   }
```
Comment (PavelLoparev) [inline Smartling.API/Authentication/AuthApiClient.cs]: Unrelated changes (spaces only?)

File: Smartling.API/Smartling.Api.csproj [hunk 4]
```diff
@@ -7,6 +7,8 @@
     <Version>2.4.34.0</Version>
     <Authors>Smartling</Authors>
     <Copyright>Copyright © Smartling 2019</Copyright>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>Key1.snk</AssemblyOriginatorKeyFile>
```
Comment (PavelLoparev) [inline Smartling.API/Smartling.Api.csproj]: Unrelated change. SDK must not know about some keys etc

File: Smartling.API/Smartling.Api.csproj [hunk 5]
```diff
@@ -15,6 +17,10 @@
     <None Remove="bin\**" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="Key1.snk" />
+  </ItemGroup>
```
Comment (PavelLoparev) [inline Smartling.API/Smartling.Api.csproj]: Unrelated change. the same

File: Smartling.API/Model/TranslationJob.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using Newtonsoft.Json;
+using System;
+
+namespace Smartling.Api.Model
+{
+    public class TranslationJob
```
Comment (PavelLoparev) [inline Smartling.API/Model/TranslationJob.cs]: Seems like we already have Job model: https://github.com/Smartling/api-sdk-net/blob/master/Smartling.API/Model/Job.cs

How are they different?

File: Smartling.API/Model/JobItem.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        public object GracefulWindow { get; set; }
+        [JsonProperty("isActionable")]
+        public bool? IsActionable { get; set; }
+        [JsonProperty("isDeclined")]
+        public bool? IsDeclined { get; set; }
+        [JsonProperty("isContentUpdated")]
+        public bool? IsContentUpdated { get; set; }
+    }
+
+    public class Issues
```
Comment (dimitrystd) [inline Smartling.API/Model/JobItem.cs]: Rest classes are placed as separate files. Why these two classes defined here?

File: Smartling.API/Dashboard/DashboardApiClient.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        private readonly IAuthenticationStrategy _auth;
+        private const string DashboardUrl = "/translator-dashboard-api/v2/jobs";
+
+        public DashboardApiClient(IAuthenticationStrategy auth)
+        {
+            _auth = auth;
+        }
+
+        public virtual DashboardData GetDashboardData()
```
Comment (dimitrystd) [inline Smartling.API/Dashboard/DashboardApiClient.cs]: Do we override it? mock in tests? Then let's don't mark it `override` if we don't have such plans

File: Smartling.API/Dashboard/DashboardApiClient.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            var uriBuilder = GetRequestStringBuilder(DashboardUrl);
+            var request = PrepareGetRequest(uriBuilder.ToString(), _auth.GetToken());
+            JObject response;
+
+            try
+            {
+                response = JObject.Parse(GetResponse(request));
```
Comment (dimitrystd) [inline Smartling.API/Dashboard/DashboardApiClient.cs]: Why not to use more high-level method `ExecuteGetRequest`?

File: Smartling.API/Model/JobItem.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
+﻿using Newtonsoft.Json;
+
+namespace Smartling.Api.Model
+{
+    public class JobItem
```
Comment (dimitrystd) [inline Smartling.API/Model/JobItem.cs]: This name confuses because of we have Jobs API and everyone who will use this SDK will be sure that this class is related to that API. I don't have better name and can only propose use another namespace.
* Put this and other "response" related files into `...Model.TranslationDashboard` folder
* Update their namespaces

File: Smartling.API/Dashboard/DashboardApiClient.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
+﻿using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Smartling.Api.Authentication;
+using Smartling.Api.Model;
+using System.Security.Authentication;
+
+namespace Smartling.Api.Dashboard
+{
+    public class DashboardApiClient : ApiClientBase
```
[the remark this answers is not in the document]
Comment (dimitrystd) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: And change folder and namespace to `TranslatorDashboard`

File: Smartling.API/Model/Account.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿using Newtonsoft.Json;
+
+namespace Smartling.Api.Model
+{
+    public class Account
+    {
+        [JsonProperty("accountName")]
```
Comment (dimitrystd) [inline Smartling.API/Model/Account.cs]: All these similar attributes are annoying. Just configure json lib to use camelCase strategy. It's a common feature in all frameworks.

File: Smartling.API/Model/TranslationJob.cs (hunk 6 above)
Comment (dimitrystd) [reply Smartling.API/Model/TranslationJob.cs]: right, duplicate

File: Smartling.APITests/AuthApiClientTests.cs [hunk 13]
```diff
@@ -6,55 +6,70 @@
...
+          // Act
+          var result = client.Object.Authenticate();
+
+          // Assert
+          Assert.AreEqual("{access token}", result.data.accessToken);
+          Assert.AreEqual("{refresh token}", result.data.refreshToken);
+        }
+
+        [TestMethod]
+        public void UserAuthenticateTest()
```
Comment (dimitrystd) [inline Smartling.APITests/AuthApiClientTests.cs]: Put your tests in separate file. 1 class file = 1 test file

File: Smartling.API/Dashboard/DashboardApiClient.cs (hunk 11 above)
Comment (dimitrystd) [inline Smartling.API/Dashboard/DashboardApiClient.cs]: Didn't see tests for this class

File: Smartling.API/Authentication/AuthApiUser.cs (hunk 2 above)
Comment (nick-ser) [reply Smartling.API/Authentication/AuthApiUser.cs]: Tested. I've got "The remote server returned an error: (401) Unauthorized."

Comment (nick-ser) [reply Smartling.API/Authentication/AuthApiUser.cs]: Yes, you are right

File: Smartling.API/Smartling.Api.csproj (hunk 4 above)
Comment (nick-ser) [reply Smartling.API/Smartling.Api.csproj]: Remove

File: Smartling.API/Smartling.Api.csproj (hunk 5 above)
Comment (nick-ser) [reply Smartling.API/Smartling.Api.csproj]: Removed

File: Smartling.API/Authentication/AuthApiClient.cs (hunk 3 above)
Comment (nick-ser) [reply Smartling.API/Authentication/AuthApiClient.cs]: There was an incorrect tab

File: Smartling.API/Model/TranslationJob.cs (hunk 6 above)
Comment (nick-ser) [reply Smartling.API/Model/TranslationJob.cs]: Removed

File: Smartling.API/Dashboard/DashboardApiClient.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (nick-ser) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: Done

File: Smartling.API/Dashboard/DashboardApiClient.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+            var request = PrepareGetRequest(uriBuilder.ToString(), _auth.GetToken());
+            JObject response;
+
+            try
+            {
+                response = JObject.Parse(GetResponse(request));
+            }
+            catch (AuthenticationException)
+            {
+                request = PrepareGetRequest(uriBuilder.ToString(), _auth.GetToken(true));
```
[the remark this answers is not in the document]
Comment (nick-ser) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: Refactored.

File: Smartling.API/Dashboard/DashboardApiClient.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+                response = JObject.Parse(GetResponse(request));
```
[the remark this answers is not in the document]
Comment (nick-ser) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: Yes, you are right. Now, this piece of code throws AuthenticationException for the executing code. The problem is that all the rest of the code works in such a manner.

File: Smartling.API/Model/JobItem.cs (hunk 7 above)
Comment (nick-ser) [reply Smartling.API/Model/JobItem.cs]: Replaced to separate files

File: Smartling.API/Dashboard/DashboardApiClient.cs (hunk 9 above)
Comment (nick-ser) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: Why is this method more higher-level if it uses GetResponse()? Also, it marks "Retry the request in case of authentication error" and all other code in this situation uses GetResponse().

File: Smartling.API/Dashboard/DashboardApiClient.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (dotnetrocks) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: > Maybe i forgot some special use cases, but if i use wrong credentials and receive 401 then second request will return the same error.

_aut.GetToken(true) within catch{} will get a fresh token

File: Smartling.API/Model/JobItem.cs (hunk 10 above)
Comment (nick-ser) [reply Smartling.API/Model/JobItem.cs]: Done.

File: Smartling.APITests/AuthApiClientTests.cs (hunk 13 above)
Comment (nick-ser) [reply Smartling.APITests/AuthApiClientTests.cs]: Done.

File: Smartling.API/Model/Account.cs (hunk 12 above)
Comment (nick-ser) [reply Smartling.API/Model/Account.cs]: Named in accordance with a codestyle used in the project.

File: Smartling.API/Dashboard/DashboardApiClient.cs (hunk 11 above)
Comment (nick-ser) [reply Smartling.API/Dashboard/DashboardApiClient.cs]: Done.

File: Smartling.API/Authentication/AuthApiUser.cs [hunk 16]
```diff
@@ -1,38 +1,12 @@
...
+﻿namespace Smartling.Api.Authentication
 {
     public class AuthApiUser : AuthApiClient
     {
-        private const string AUTH_API_V2_AUTHENTICATE_USER = "/auth-api/v2/authenticate/user";
-        private const string AUTH_API_V2_REFRESH_USER = "/auth-api/v2/authenticate/refresh/user";
-        private readonly string _userIdentifier;
-        private readonly string _userSecret;
+        protected override string AuthApiV2Authenticate { get { return "/auth-api/v2/authenticate/user"; } }
+        protected override string AuthApiV2Refresh { get { return "/auth-api/v2/authenticate/refresh/user"; } }
```
Comment (PavelLoparev) [inline Smartling.API/Authentication/AuthApiUser.cs]: Again, I think refresh uri should be `/auth-api/v2/authenticate/refresh`

File: Smartling.API/Authentication/AuthApiUser.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Smartling.Api.Authentication
+{
+    public class AuthApiUser : AuthApiClient
+    {
+        protected override string AuthApiV2Authenticate { get { return "/auth-api/v2/authenticate/user"; } }
+
+        public AuthApiUser(string userIdentifier, string userSecret) : base(userIdentifier, userSecret)
+        {
+        }
```
Comment (PavelLoparev) [inline Smartling.API/Authentication/AuthApiUser.cs]: I don't know C#, just a question. Why did you override define this method? Seems like it has an empty body.

File: Smartling.API/Smartling.Api.csproj [hunk 18]
```diff
@@ -7,6 +7,9 @@
     <Version>2.4.34.0</Version>
     <Authors>Smartling</Authors>
     <Copyright>Copyright © Smartling 2019</Copyright>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>key1.snk</AssemblyOriginatorKeyFile>
+    <DelaySign>false</DelaySign>
```
Comment (PavelLoparev) [inline Smartling.API/Smartling.Api.csproj]: Once again - **sdk must not know about some keys** etc. Can you just open this code not in `trados plugin` solution but as a separate project? I think it's your ide generates this.

**Comment by dimitrystd** [comment]
Close obsolete PR without merge
